airflow/models.py

```diff
@@ -3073,6 +3074,53 @@ def cli(self):
         args = parser.parse_args()
         args.func(args, self)
 
+    @provide_session
+    def create_dagrun(self,
+                      run_id,
+                      execution_date,
+                      state,
+                      start_date=None,
+                      external_trigger=False,
+                      conf=None,
+                      session=None):
+        """
+        Creates a dag run from this dag including the tasks associated with this dag. Returns the dag
+        run.
+        :param run_id: defines the the run id for this dag run
+        :type run_id: string
+        :param execution_date: the execution date of this dag run
+        :type execution_date: datetime
+        :param start_date: the date this dag run should be evaluated
+        :type state_date: datetime
+        :param state: the state of the dag run
+        :type state: State
+        :param external_trigger: whether this dag run is externally triggered
+        :type external_trigger: bool
+        """
+        run = DagRun(
+            dag_id=self.dag_id,
+            run_id=run_id,
+            execution_date=execution_date,
+            start_date=start_date,
+            external_trigger=external_trigger,
+            conf=conf,
+            state=state
+        )
+        session.add(run)
+
+        for task in self.tasks:
+            # todo: check this
+            if task.adhoc:
+                continue
+
+            ti = TaskInstance(task, execution_date)
+            session.add(ti)
+
+        session.commit()
+
+        run.refresh_from_db()
```

Comment:
Ah the important bit: this create_dagrun creates the *full* dagrun and thus includes creating the taskinstances for this dagrun. Considering that a Dag is an ordered combination of tasks a DagRun should also be this at any given point in time. Ie. the current situation violates this: DagRuns are created shallow, without TaskInstances.

Having this in resolves issues with some occurrences of deadlocks and depend_on_past and I do consider it important for the integrity of airflow going forward and is one of the reasons of using "create_dagrun" to make sure we will be consistent in creating dag runs.